// @percy/config/dist/migrate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addMigration = addMigration;
exports.clearMigrations = clearMigrations;
exports.default = migrate;

var _logger = _interopRequireDefault(require("@percy/logger"));

var _utils = require("./utils");

var _normalize = _interopRequireDefault(require("./normalize"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Global set of registered migrations
const migrations = new Set(); // Register a migration function

function addMigration(migration) {
  migrations.add(migration);
} // Clear all migration functions


function clearMigrations() {
  migrations.clear();
} // Calls each registered migration function with a normalize provided config
// and util functions for working with the config object


function migrate(config) {
  config = (0, _normalize.default)(config);
  let util = {
    set: _utils.set.bind(null, config),
    map: _utils.map.bind(null, config),
    del: _utils.del.bind(null, config),
    log: (0, _logger.default)('config')
  };

  for (let migration of migrations) {
    migration(config, util);
  }

  return (0, _utils.merge)([config, {
    version: 2
  }]);
}
